File: Semester 03 Projects/Solitair Game/BL/Queue.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Solitair_Game
{
    //Queue Class implementing queue using Linked List.
    public class Queue
    {
        public Node Front;
        public Node Rear;
        public Queue()
        {
            Front = null;
            Rear= null;
        }

        //Function for making Deep Copy of Queue.
        public Queue DeepCopy()
        {
            Queue copy = new Queue();
            if (Front == null) {
                return copy;
            }
            Queue tempQueue = new Queue();
            Node current = Front;
            while(current!=null)
            {
                Card copycard = new Card(current.CurrentCard);
                tempQueue.Enqueue(copycard);
                current = current.Next;
            }
            current = tempQueue.Front;
            while(current!=null)
            {
                copy.Enqueue(current.CurrentCard);
                current = current.Next;
            }
            return copy;
        }

        //Function for adding card into queue.
        public void Enqueue(Card newcard)
        {
            Node newNode = new Node();
            newNode.CurrentCard = newcard;
            newNode.Next = null;
            if (Front == null)
            {
                Front = newNode;
                Rear= newNode;
            }
            else
            {
                Rear.Next = newNode;
                Rear = newNode;
            }
        }

        //Function for removing card from queue.
        public Card Dequeue()
        {
            if (IsEmpty())
            {
                return null;
            }
            Node next = Front.Next;
            Card CardToDelete = Front.CurrentCard;
            Front = next;
            if (Front == null)
            {
                Rear = null;
            }
            return CardToDelete;
        }

        //Function for returning front of queue.
        public Card Peek()
        {
            if (IsEmpty())
            {
                return null;
            }
            return Front.CurrentCard;
        }

        //Function for checking queue is empty or not.
        public bool IsEmpty()
        {
            if (Front == null)
            {
                return true;
            }
            return false;
        }

        //Function for counting number of cards in queue.
        public int GetTotalNumberOfCards()
        {
            int count = 0;
            Node current = Front;
            while (current != null)
            {
                count++;
                current = current.Next;
            }
            return count;
        }
    }
}

File: DSA-Final-Project/Server/DataStructuresAndAlgorithms/Queue.js
const fs = require("fs");
class Node 
{
    constructor(value) 
    {
        this.value = value;
        this.next = null;
    }
}

class Queue 
{
    constructor() 
    {
        this.head = null;
        this.tail = null;
    }

    Enqueue(value) 
    {
        const newNode = new Node(value);
        if (this.head === null) 
        {
            this.head = newNode;
            this.tail = newNode;
        } 
        else 
        {
            this.tail.next = newNode;
            this.tail = newNode;
        }
    }

    Dequeue() 
    {
        if (this.IsEmpty()) 
        {
            return "No elements to dequeue";
        }
        const dequeuedValue = this.head.value;
        this.head = this.head.next;
        if (this.head === null) 
        {
            this.tail = null;
        }
        return dequeuedValue;
    }

    Peek() 
    {
        if (this.IsEmpty()) 
        {
            return "Queue is empty";
        }
        return this.head.value;
    }

    IsEmpty() 
    {
        if(this.head === null)
        {
            return true
        }
        return false;
    }
    DeleteSpecific(location) {
        if (this.head == null) {
            return false;
        }

        if (this.head.value.UserID === location.UserID &&
            this.head.value.SourceLocation === location.SourceLocation &&
            this.head.value.DestinationLocation === location.DestinationLocation) 
        {
            const locationToDelete = this.head;
            this.head = this.head.next; 
            if (this.head == null) 
            {
                this.tail = null; 
            }
            return locationToDelete.value;
        }

        let current = this.head;
        while (current.next != null) 
            {
            if (current.next.value.UserID === location.UserID &&
                current.next.value.SourceLocation === location.SourceLocation &&
                current.next.value.DestinationLocation === location.DestinationLocation)
                {
                const locationToDelete = current.next;
                if (current.next.next != null) 
                {
                    current.next = current.next.next; 
                } 
                else 
                {
                    current.next = null; 
                    this.tail = current;
                }
                return locationToDelete.value;
            }
            current = current.next;
        }
    
        return false; 
    }
}
    

module.exports = Queue;